### PR TITLE
[NOD-1579] Improve the IBD mechanism

### PR DESCRIPTION
**Pull request by stasatdaglabs**

### CODE REVIEW

File: domain/consensus/processes/blockprocessor/validateandinsertblock.go [hunk 1]
```diff
@@ -11,88 +11,51 @@ import (
 	"github.com/pkg/errors"
 )
 
+type insertMode uint8
+
+const (
+	insertModeGenesis insertMode = iota
+	insertModeHeader
+	insertModeBlockWithoutUpdatingVirtual
```
Comment (someone235) [inline domain/consensus/processes/blockprocessor/validateandinsertblock.go:19]: Isn't insertMissingBlockBody a better name?

File: domain/consensus/processes/blockprocessor/validateandinsertblock.go [hunk 2]
```diff
@@ -179,6 +137,62 @@ func (bp *blockProcessor) validateAndInsertBlock(block *externalapi.DomainBlock)
...
+	if syncState == externalapi.SyncStateAwaitingBlockBodies {
+		headerTips, err := bp.headerTipsStore.Tips(bp.databaseContext)
+		if err != nil {
+			return 0, err
+		}
+		selectedHeaderTip, err := bp.ghostdagManager.ChooseSelectedParent(headerTips...)
+		if err != nil {
+			return 0, err
+		}
+		if *selectedHeaderTip != *blockHash {
```
Comment (someone235) [inline domain/consensus/processes/blockprocessor/validateandinsertblock.go:184]: This condition is redundant

File: domain/consensus/processes/blockprocessor/validateandinsertblock.go [hunk 3]
```diff
@@ -215,14 +224,15 @@ func (bp *blockProcessor) checkBlockStatus(hash *externalapi.DomainHash, mode *e
 		return errors.Wrapf(ruleerrors.ErrKnownInvalid, "block %s is a known invalid block", hash)
 	}
 
-	if mode.State == externalapi.SyncStateHeadersFirst || status != externalapi.StatusHeaderOnly {
+	isBlockBodyAfterBlockHeader := mode != insertModeHeader && status == externalapi.StatusHeaderOnly
+	if !isBlockBodyAfterBlockHeader {
```
Comment (someone235) [inline domain/consensus/processes/blockprocessor/validateandinsertblock.go:228]: What about duplicate header?

File: domain/consensus/processes/blockprocessor/validateandinsertblock.go [hunk 4]
```diff
@@ -283,12 +293,18 @@ func (bp *blockProcessor) validatePreProofOfWork(block *externalapi.DomainBlock)
 	return nil
 }
 
-func (bp *blockProcessor) validatePostProofOfWork(block *externalapi.DomainBlock, mode *externalapi.SyncInfo) error {
-	blockHash := consensushashing.BlockHash(block)
-
-	if mode.State != externalapi.SyncStateHeadersFirst {
+func (bp *blockProcessor) validatePruningPointViolationAndProofOfWorkAndDifficulty(block *externalapi.DomainBlock, mode insertMode) error {
+	blockHash := consensushashing.HeaderHash(block.Header)
+	if mode != insertModeHeader {
```
Comment (someone235) [inline domain/consensus/processes/blockprocessor/validateandinsertblock.go:298]: Please add a comment why you do this

File: domain/consensus/processes/syncmanager/blocklocator.go [hunk 5]
```diff
@@ -28,6 +28,11 @@ func (sm *syncManager) createBlockLocator(lowHash, highHash *externalapi.DomainH
 	for currentHash != nil {
 		locator = append(locator, currentHash)
 
+		// Stop if we've reached the limit (if it's set)
+		if limit != nil && len(locator) >= *limit {
```
Comment (someone235) [inline domain/consensus/processes/syncmanager/blocklocator.go:32]: Can you make it == limit

**Review by someone235 — CHANGES_REQUESTED** [review]

File: domain/consensus/processes/syncmanager/antipast.go [hunk 6]
```diff
@@ -110,60 +110,41 @@ func (sm *syncManager) missingBlockBodyHashes(highHash *externalapi.DomainHash)
...
+			foundHeaderOnlyBlock = true
 			break
 		}
+		lowHash = selectedChild
 	}
-
-	hashesBetween, err := sm.antiPastHashesBetween(lowHash, highHash)
-	if err != nil {
-		return nil, err
+	if !foundHeaderOnlyBlock {
```
Comment (someone235) [inline domain/consensus/processes/syncmanager/antipast.go:127]: Can you try to revert it to use selectedChildIterator but use reachability's children instead?

**Review by someone235 — CHANGES_REQUESTED** [review]

File: domain/consensus/processes/blockprocessor/validateandinsertblock.go (hunk 2 above)
Comment (stasatdaglabs) [reply domain/consensus/processes/blockprocessor/validateandinsertblock.go:184]: We spoke about this and decided to leave it as is.

File: domain/consensus/processes/syncmanager/antipast.go (hunk 6 above)
Comment (stasatdaglabs) [reply domain/consensus/processes/syncmanager/antipast.go:127]: I tried to, but unfortunately, I still got wrong results.

File: app/protocol/flows/blockrelay/handle_relay_invs.go [hunk 7]
```diff
@@ -211,52 +198,98 @@ func (flow *handleRelayInvsFlow) readMsgBlock() (msgBlock *appmessage.MsgBlock,
...
-	err = flow.Broadcast(appmessage.NewMsgInvBlock(blockHash))
+func (flow *handleRelayInvsFlow) relayBlock(block *externalapi.DomainBlock) error {
+	blockHash := consensushashing.BlockHash(block)
+	err := flow.Broadcast(appmessage.NewMsgInvBlock(blockHash))
 	if err != nil {
 		return err
 	}
 
-	log.Debugf("Accepted block %s via relay", blockHash)
+	log.Infof("Accepted block %s via relay", blockHash)
```
Comment (someone235) [inline app/protocol/flows/blockrelay/handle_relay_invs.go]: Can you remove this log?

**Review by someone235 — CHANGES_REQUESTED** [review]

**Review by someone235 — APPROVED** [review]

**Pull request merged**